### ISSUE TITLE
Support more granular timeouts for queries

### ISSUE THREAD

**Issue by matthax** [ClickHouse/clickhouse-connect#192]
**Is your feature request related to a problem? Please describe.**
Feature request / enhancement

**Describe the solution you'd like**
I'd like to have more granular control over timeouts in the client. Specifically, the `ping` method has a [hardcoded timeout](https://github.com/ClickHouse/clickhouse-connect/blob/fa20547d7f7e2fd3a2cf4cd711c3262c5a79be7a/clickhouse_connect/driver/httpclient.py#L414) of `3` but in our use case we'd like to be able to configure that to something like `1` for example.
 - The `clickhouse-driver` library [solves this](https://clickhouse-driver.readthedocs.io/en/latest/api.html#connection) using a `sync_request_timeout` argument. Perhaps something similar could be added to the client constructor here.

The `_raw_request` method [uses the timeout](https://github.com/ClickHouse/clickhouse-connect/blob/fa20547d7f7e2fd3a2cf4cd711c3262c5a79be7a/clickhouse_connect/driver/httpclient.py#L361) on the client instance, which is more flexible, but our use case requires a per-query timeout setting. Currently, I'm working around this with a solution I don't love

```python3
    @contextmanager
    def with_timeout(
        self,
        connect_timeout: Optional[int] = None,
        send_receive_timeout: Optional[int] = None,
    ):
        _old_timeout = self._delegate.timeout
        self._delegate.timeout = Timeout(
            connect=coerce_int(connect_timeout or _old_timeout.connect_timeout),
            read=coerce_int(send_receive_timeout or _old_timeout.read_timeout),
        )
        yield self._delegate
        self._delegate.timeout = _old_timeout
```

This approach is similar to that in `clickhouse-driver` which offers a `timeout_setter` on the `Connection` instance it provides. A builtin way to supply a timeout either through a context manager or with a sort of per-query setting e.g. `client.query(..., connect_timeout=1, send_receive_timeout=3)` would be great.

**Describe alternatives you've considered**
Right now I'm composing a class that adds some workarounds (the `with_timeout` above) and for ping I override the method
```python3
    def ping(self):
        """
        Ping the host, sync_request_timeout is supplied to the call
        TODO: Suggest this as a PR to clickhouse-connect
        """
        try:
            response = self._delegate.http.request(
                "GET", f"{self._delegate.url}/ping", timeout=self._sync_request_timeout
            )
            return 200 <= response.status < 300
        except HTTPError:
            return False
```

I suppose query context or settings could also be a feasible way to supply the timeout. I don't have a strong opinion on how it should be accomplished and would love to hear your thoughts.

**Additional context**

I'm happy to PR any of these changes, just wanted to start a discussion on the approach before starting any implementation.

**Comment by matthax** [comment]
@genzgd I definitely hear you on the argument overload. For the provider option, what would you envision that function receiving as arguments?

Adding a default to ping would be great, that's the bigger pain point of the two. 

And yes, the primary concern I had with the approach I demonstrated above was thread safety, but we do have locks around the client connection pulling that should prevent issues there.

**Comment by genzgd** [comment]
@matthax  Thinking through this if you really need per query timeouts they should be added into the QueryContext (even though it has so many properties already).  My provider function idea is just a little too fancy.

Since it's an advanced feature I think using a `union` of float/int/urllib3 Timeout should be okay, since it will just get passed to the urllib3 library anyway.  Any of the methods (inserts, commands, raw*) that need such a timeout should accept that additional parameter.

If you want to take a stab at a PR along those lines that would be most welcome, otherwise it will probably be a week or so before I get to it (in connection with a few other enhancements I'm working on as I have time.)

**Comment by matthax** [comment]
@genzgd I can open a PR for this

**Comment by joe-clickhouse** [comment]
Hey @matthax thanks for the suggestion. I know it's been a while...do you still have any interest in this or a need/use case for it?

**Comment by matthax** [comment]
> Hey [@matthax](https://github.com/matthax) thanks for the suggestion. I know it's been a while...do you still have any interest in this or a need/use case for it?


I have been monkey patching this for ~3y, it would still be nice to have, but I was not approved to open the MR.

**Comment by joe-clickhouse** [comment]
@matthax haha I hear ya. I can add e.g. `ping(timeout=1.0)`. Would that let you remove the monkey patch? Or do you also still change connect/read timeouts per query on the same client? If so, could you share one current example?
